In [ ]:
#!pip install tensorflow

In [1]:
import tensorflow
tensorflow.__version__

'2.3.1'

In [2]:
import warnings
warnings.filterwarnings("ignore")

In [3]:
#Importing the dataset
from tensorflow.keras.datasets import boston_housing


In [ ]:
# when you use dataset.loaddata() : By default the data is split into 80 % and 20% randomly
# and it will return 2 tuples
# tuple 1  X_train , y_train (80 % of the data)
#tuple 2 X_test , y_test (20 % of the data)

In [5]:
(features, prices ), _  = boston_housing.load_data(test_split=0)

In [6]:
print("No of examples:",features.shape[0])

No of examples: 506


In [7]:
print("no of columns",features.shape[1])

no of columns 13


In [8]:
import pandas as pd

In [9]:
sample = pd.DataFrame(features)

In [10]:
sample.head()

0     1      2    3      4      5      6       7     8      9    10  \
0  1.23247   0.0   8.14  0.0  0.538  6.142   91.7  3.9769   4.0  307.0  21.0   
1  0.02177  82.5   2.03  0.0  0.415  7.610   15.7  6.2700   2.0  348.0  14.7   
2  4.89822   0.0  18.10  0.0  0.631  4.970  100.0  1.3325  24.0  666.0  20.2   
3  0.03961   0.0   5.19  0.0  0.515  6.037   34.5  5.9853   5.0  224.0  20.2   
4  3.69311   0.0  18.10  0.0  0.713  6.376   88.4  2.5671  24.0  666.0  20.2   

       11     12  
0  396.90  18.72  
1  395.38   3.11  
2  375.52   3.26  
3  396.90   8.01  
4  391.43  14.65

- My dataset contains 506 rows and 13 columns and I am trying to predict a house price based on the independent variables.

- how many neurons in my input layer - ?  ( 13) 

- what is the loss function -?  ( mean squared ) 

- how many neurons in my last layer -?  (1)

- Sequential API

- when we did our playground, is it not sequential? we had our input layer then we added one hidden layer and one hidden layer and last layer 



#### Sequential API
- Sequence is a linear stack of layers
- The model should know what input shape it is taking




In [23]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import BatchNormalization , Dense

In [31]:
model = Sequential()

In [32]:
model.add(Dense(5,input_shape=(13,))) #input layer with 13 neurons and 10 neurons on hidden
model.add(Dense(1)) #prediction layer
model.compile(optimizer="sgd",loss="mse")
print(model.summary())

Model: "sequential_4"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_12 (Dense)             (None, 5)                 70        
_________________________________________________________________
dense_13 (Dense)             (None, 1)                 6         
Total params: 76
Trainable params: 76
Non-trainable params: 0
_________________________________________________________________
None


In [33]:
model.fit(features,prices,epochs=100,validation_split=0.30)

Epoch 1/100
12/12 [==============================] - 0s 7ms/step - loss: nan - val_loss: nan
Epoch 2/100
12/12 [==============================] - 0s 2ms/step - loss: nan - val_loss: nan
Epoch 3/100
12/12 [==============================] - 0s 2ms/step - loss: nan - val_loss: nan
Epoch 4/100
12/12 [==============================] - 0s 2ms/step - loss: nan - val_loss: nan
Epoch 5/100
12/12 [==============================] - 0s 2ms/step - loss: nan - val_loss: nan
Epoch 6/100
12/12 [==============================] - 0s 2ms/step - loss: nan - val_loss: nan
Epoch 7/100
12/12 [==============================] - 0s 2ms/step - loss: nan - val_loss: nan
Epoch 8/100
12/12 [==============================] - 0s 2ms/step - loss: nan - val_loss: nan
Epoch 9/100
12/12 [==============================] - 0s 2ms/step - loss: nan - val_loss: nan
Epoch 10/100
12/12 [==============================] - 0s 3ms/step - loss: nan - val_loss: nan
Epoch 11/100
12/12 [==============================] - 0s 3ms/step - l

12/12 [==============================] - 0s 2ms/step - loss: nan - val_loss: nan
Epoch 89/100
12/12 [==============================] - 0s 3ms/step - loss: nan - val_loss: nan
Epoch 90/100
12/12 [==============================] - 0s 2ms/step - loss: nan - val_loss: nan
Epoch 91/100
12/12 [==============================] - 0s 2ms/step - loss: nan - val_loss: nan
Epoch 92/100
12/12 [==============================] - 0s 3ms/step - loss: nan - val_loss: nan
Epoch 93/100
12/12 [==============================] - 0s 2ms/step - loss: nan - val_loss: nan
Epoch 94/100
12/12 [==============================] - 0s 2ms/step - loss: nan - val_loss: nan
Epoch 95/100
12/12 [==============================] - 0s 2ms/step - loss: nan - val_loss: nan
Epoch 96/100
12/12 [==============================] - 0s 2ms/step - loss: nan - val_loss: nan
Epoch 97/100
12/12 [==============================] - 0s 2ms/step - loss: nan - val_loss: nan
Epoch 98/100
12/12 [==============================] - 0s 2ms/step - loss:

In [13]:
# #Dense is a keyword that represents a fully connected layer
# model.add(BatchNormalization(input_shape=(13,)))
# model.add(Dense(1))

In [ ]:
#sgd  = Stociastic gradient descent
#mse = mean squared error

Epoch 1/100
12/12 [==============================] - 0s 10ms/step - loss: 260.3403 - val_loss: 210.2843
Epoch 2/100
12/12 [==============================] - 0s 2ms/step - loss: 64.9502 - val_loss: 105.9042
Epoch 3/100
12/12 [==============================] - 0s 2ms/step - loss: 50.0637 - val_loss: 107.6365
Epoch 4/100
12/12 [==============================] - 0s 2ms/step - loss: 52.6773 - val_loss: 486.2982
Epoch 5/100
12/12 [==============================] - 0s 3ms/step - loss: 35.1120 - val_loss: 327.6465
Epoch 6/100
12/12 [==============================] - 0s 3ms/step - loss: 37.4038 - val_loss: 101.3918
Epoch 7/100
12/12 [==============================] - 0s 3ms/step - loss: 31.3483 - val_loss: 79.4592
Epoch 8/100
12/12 [==============================] - 0s 3ms/step - loss: 28.5949 - val_loss: 31.7630
Epoch 9/100
12/12 [==============================] - 0s 3ms/step - loss: 31.1658 - val_loss: 83.1333
Epoch 10/100
12/12 [==============================] - 0s 3ms/step - loss: 31.2314 -

12/12 [==============================] - 0s 2ms/step - loss: 31.8443 - val_loss: 44.2939
Epoch 82/100
12/12 [==============================] - 0s 2ms/step - loss: 27.5073 - val_loss: 108.0843
Epoch 83/100
12/12 [==============================] - 0s 2ms/step - loss: 37.0788 - val_loss: 42.5816
Epoch 84/100
12/12 [==============================] - 0s 3ms/step - loss: 30.6545 - val_loss: 29.1259
Epoch 85/100
12/12 [==============================] - 0s 3ms/step - loss: 30.3099 - val_loss: 44.6113
Epoch 86/100
12/12 [==============================] - 0s 2ms/step - loss: 33.7082 - val_loss: 29.1503
Epoch 87/100
12/12 [==============================] - 0s 3ms/step - loss: 32.4824 - val_loss: 80.1526
Epoch 88/100
12/12 [==============================] - 0s 2ms/step - loss: 31.5035 - val_loss: 29.2445
Epoch 89/100
12/12 [==============================] - 0s 2ms/step - loss: 27.0750 - val_loss: 28.4652
Epoch 90/100
12/12 [==============================] - 0s 2ms/step - loss: 34.8963 - val_loss: 